pimd/pim_register.c

```diff
@@ -425,6 +488,44 @@ int pim_register_recv(struct interface *ifp, pim_addr dest_addr,
 	memset(&sg, 0, sizeof(sg));
 	sg = pim_sgaddr_from_iphdr(ip_hdr);
 
+#if PIM_IPV == 6
+	/*
+	 * According to RFC section 4.9.3, If Dummy PIM Header is included 
+         * in NULL Register the dummy PIM Header's checksum should 
+         * be validated in addition to the outer PIM Header's checksum
+	 */
+	if ((*bits & PIM_REGISTER_NR_BIT) &&
+	    ((tlv_buf_size - PIM_MSG_REGISTER_BIT_RESERVED_LEN) >
+	     (int)sizeof(struct ip6_hdr))) {
+		uint16_t computed_checksum;
+		uint16_t received_checksum;
+		struct ipv6_ph ph;
+		struct pim_msg_header *header;
+
+		header = (struct pim_msg_header
+				  *)(tlv_buf +
+				     PIM_MSG_REGISTER_BIT_RESERVED_LEN +
+				     sizeof(struct ip6_hdr));
+		ph.src = sg.src;
+		ph.dst = sg.grp;
+		ph.ulpl = htonl(PIM_MSG_HEADER_LEN);
+		ph.next_hdr = IPPROTO_PIM;
+
+		received_checksum = header->checksum;
+
+		header->checksum = 0;
+		computed_checksum = in_cksum_with_ph6(
+			&ph, header, htonl(PIM_MSG_HEADER_LEN));
+
+		if (computed_checksum != received_checksum) {
```

Comment:
The checksum calculation is already handled in the caller function pim_pim_packet. Can you please check it.

Reply:
According to RFC section 4.9.3, If Dummy PIM Header is included in NULL Register the dummy PIM Header's checksum should be validated in addition to the outer PIM Header's checksum
         in pim_pim_packet.c it validates the outer PIM header checksum.

  I have mentioned this in comments

Reply:
Please refer to this commit https://github.com/FRRouting/frr/pull/10882/commits/721c7be3b20a45ab99cc9e94d92a1e9d9cd5f3bb. The checksum including the dummy header is already handled in the caller pim_pim_packet.

Reply:
Please refer to this commit https://github.com/FRRouting/frr/pull/10882/commits/721c7be3b20a45ab99cc9e94d92a1e9d9cd5f3bb. The checksum including the dummy header is already handled in the caller pim_pim_packet.

Reply:
The commit you have referred, validates the checksum in outer PIM header (this is computed including full payload including dummy PIM header). Here I am checking the checksum field separately filled in inner dummy PIM header in Null registers. that also has to be validated separately.

Reply:
The discussion here shows that it is easy to misunderstand

@plsaranya please add a comment explaining that this is the inner checksum being validated not the outer, so people looking at it in the future can understand it easier.

Reply:
Ah i see you have the comment above the if block… I don't think it is clear enough, can you try to make it more explanatory?